### PR TITLE
ENH: Add support for multiple conditions assign statement

### PR DESCRIPTION
**Pull request by ELHoussineT**
Used to support conditional assignment operation.

Usage example: 
```
df.assign(d = pd.case_when(lambda x: <some conditions>, 0,            # first condition, result   
                           (other_df.z > 4) | (other_df.x.isna()), 1, # second condition, result 
                           "some_value"))                             # default (optional)
```

Continued discussion [here](https://github.com/pandas-dev/pandas/issues/39154#issuecomment-1357794823).

Credit: 
Some of the logic is inspired by [this implementation](https://github.com/pyjanitor-devs/pyjanitor/blob/cdf62dac8696aabae0946b401de2b94313c23fc1/janitor/functions/case_when.py) which at the time of constructing the PR was contributed to by @samukweku @Zeroto521 @thatlittleboy and @ericmjl 

- [x] closes #39154 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by erfannariman** [comment]
I am not sure if I would've implemented it with an anonymous function (eg. lambda). Since `np.select` expects boolean arrays, so we could just use a standard pandas condition `df["a"] == "some value"`

File: pandas/core/case_when.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    Parameters
+    ----------
+    args : Variable argument of conditions and expected values.
+        Takes the form:
+            `condition0`, `value0`, `condition1`, `value1`, ...
+        `condition` can be a 1-D boolean array/series or a callable
+        that evaluate to a 1-D boolean array/series.
+    default : Any, default is `None`.
+        The default value to be used if all conditions evaluate False.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:30]: If no value is specified, would we not want the default for it to remain untouched?

When default is not specified, I believe the current implementation will convert e.g. int to object dtypes. That seems undesirable to me.

File: pandas/core/case_when.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+import numpy as np
+
+from pandas._libs import lib
+
+import pandas as pd
+import pandas.core.common as com
+
+
+def case_when(*args, default: Any = lib.no_default) -> Callable:
```
Comment (rhshadrach) [inline pandas/core/case_when.py:16]: Users wanting to use this without assign would be calling something like

    pd.case_when(cond1, val1, cond2, val2)(df)

That seems odd, and it's much more natural to have the DataFrame as an argument. On the other hand, that wouldn't allow as much usage with assign via method chaining if the DataFrame is changed as part of the chain.

Comment (samukweku) [reply pandas/core/case_when.py:16]: I think it should return a series instead

File: pandas/core/case_when.py (hunk 1 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: Indeed. For now, we can use `np.nan`. Pushed a fixup: [5595161](https://github.com/pandas-dev/pandas/pull/50343/commits/559516128b09567b623626119f63afad19f9e05b)

Using `np.nan` also preserves date related dtypes properly. To illustrate: 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3]))
>>> result = df.assign(
...     new_column=pd.case_when(
...         lambda x: x.a == 1, pd.Timestamp("today")
...     )
... )
>>> result.new_column
0   2023-01-07 18:58:28.654455
1                          NaT
2                          NaT
Name: new_column, dtype: datetime64[ns]
```

File: pandas/core/case_when.py (hunk 2 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py:16]: > On the other hand, that wouldn't allow as much usage with assign via method chaining if the DataFrame is changed as part of the chain.

Exactly, what do you suggest in this case?

File: pandas/core/case_when.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    if default is lib.no_default:
+        default = np.nan
+
+    def _eval(df: pd.DataFrame) -> np.ndarray:
+        booleans = []
+        replacements = []
+
+        for index, value in enumerate(args):
+            if not index % 2:
+                if callable(value):
```
Comment (samukweku) [inline pandas/core/case_when.py:85]: I think you can skip the callable check

File: pandas/core/case_when.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                    value = com.apply_if_callable(value, df)
+                booleans.append(value)
+            else:
+                replacements.append(value)
+
+        return np.select(booleans, replacements, default=default)
```
Comment (samukweku) [inline pandas/core/case_when.py:91]: What happens if there is an extension dtype? I don't think numpy select handles that well

File: pandas/core/case_when.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/core/case_when.py]: I was in the mindset that we were doing something like

```
ser.mask(cond1, value1).mask(~cond1 & cond2, value2)
```

with a starting Series. You can indeed get this behavior with case_when by specifying `.case_when(..., default=ser)`. So I think behavior-wise this is good.

However I don't think the default of np.nan will work for all dtypes, e.g. in my experience it's common (but not universal) to have None instead of np.nan be the NA value for string. Leaving it as np.nan might unexpectedly coerce the dtype on some inputs as well, surprising users.

With this, it seems better to me to have `default` be a required argument.

File: pandas/core/case_when.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/core/case_when.py:16]: I don't think we should start adding functions to the API just to support method chaining in assign. Users can still use this with a simple wrapper, e.g.

```
def chain_case_when(df, *args, **kwargs):
    return lambda df: pd.case_when(df, *args, **kwargs)

result = (
    ...
    .assign(new_col=chain_case_when(lambda x: x.a == 1, 'first', lambda x: (x.a > 1), 'second', default='default'))
)
```

A few other thoughts on the API. A few of these repeat other comments here, just adding so it's all in one place.

 - I think the `default` argument should have no default (must be user-specified) https://github.com/pandas-dev/pandas/pull/50343#discussion_r1064136418
 - Return value should be pandas object (Series)
 - The top level method is necessary for constructing a Series where the default is a scalar. However it also seems natural to also have `Series.case_when` where `default` is the provided Series (so not an argument in the signature). We could also have `DataFrame.case_when` be an alias for `pd.case_when(df, ...)`.
 - Needs to support EAs (https://github.com/pandas-dev/pandas/pull/50343#discussion_r1064051279)
 - Does the current implementation support replacements (values) being Series? I think the answer is yes, but wanted to be sure. This should be added to docs and tests.

The tests should also be expanded:
 - condition being the Boolean EA
 - dtypes for values including bool, string, and EAs (numeric, categorical, datetimelike, period)
 - Malformed arguments. E.g. what happens when a condition is something unexpected like a scalar or a list / ndarray / Series that is the wrong length or not aligned?

Comment (samukweku) [reply pandas/core/case_when.py:16]: I personally feel pd.Series.mask should be used instead of np.select. the implementation of pd.Series.mask takes care of a lot of things, including dtypes and alignment. 

Having case_when as a function makes it versatile, one can use it within assign or on its own. Agreed that it should return a pandas object, preferably a series.

Comment (rhshadrach) [reply pandas/core/case_when.py:16]: Agreed on using mask (or mask-like internals if there is unnecessary logic that is wasting cycles)

> Having case_when as a function makes it versatile

Just for clarification, is this opposition to also having `Series.case_when`?

Comment (samukweku) [reply pandas/core/case_when.py:16]: Since pd.case_when returns a pandas object I don't see a need for series.case_when( although it is convenient).

Comment (rhshadrach) [reply pandas/core/case_when.py:16]: Agreed it's not logically necessary, but it seems to be a common use case and using the a top level function is unnecessarily redundant:

    # If case_when supports Series, then can remove the `.to_frame()`
    pd.case_when(ser.to_frame(), cond1, val1, cond2, val2, default=ser)

as opposed to

    ser.case_when(cond1, val1, cond2, val2)

**Comment by samukweku** [comment]
@ELHoussineT how's it going?

**Comment by ELHoussineT** [comment]
Sorry I was away. Will push commits this weekend. @samukweku

**Comment by ELHoussineT** [comment]
@rhshadrach @samukweku 

Hey guys, sorry for the late reply. Moving forward, I should be able to address this PR quickly until we hopefully merge it. 

Thanks a lot for the valuable feedback. 

Since the feedback contained core changed to the logic, let's first agree that [0e053ef](https://github.com/pandas-dev/pandas/pull/50343/commits/0e053ef03be0c9e3e92199bb2748a2967b70c73d) is how we desire the API to behave. Once agreed, I will proceed with adjusting the tests, docs, ...etc. 

As you can see in the last commit, the following changes are in effect: 
1. `default` is required 
2. The API returns a `Series`
3. The API uses `Series.mask` under the hood which give us a lot of advantages specially dealing with EAs.

If you agree on that logic, I will proceed with the tests & docs; otherwise, we can further iterate on the logic until it meets our desired. 

Looking forward to your response.

File: pandas/core/case_when.py (hunk 4 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: Kindly see https://github.com/pandas-dev/pandas/pull/50343#issuecomment-1460350497

File: pandas/core/case_when.py (hunk 3 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: With the switch to `Series.mask` (see the switch [here](https://github.com/pandas-dev/pandas/pull/50343#issuecomment-1460350497)), this is necessary to support passing conditions to `case_when` as a callable.

File: pandas/core/case_when.py (hunk 2 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: Kindly see https://github.com/pandas-dev/pandas/pull/50343#issuecomment-1460350497

File: pandas/core/case_when.py (hunk 1 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: > With this, it seems better to me to have default be a required argument.

Good idea, done. 

> Also, how does the performance compare between:

Is there a need to check this performance as now `case_when` uses `Series.mask`?

File: pandas/core/case_when.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    Uses `Series.mask` to perform the assignment.
+
+    Parameters
+    ----------
+    obj : Dataframe or Series on which the conditions will be applied.
+    args : Variable argument of conditions and expected values.
+        Takes the form:
+            `condition0`, `value0`, `condition1`, `value1`, ...
+        `condition` can be a 1-D boolean array/series or a callable
+        that evaluate to a 1-D boolean array/series. See examples below.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:25]: value0 et al can be Series here, yes?

File: pandas/core/case_when.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        will be used to create the `Series` on which `Series.mask` will be called.
+        If this value is not already an array like (i.e. it is not of type `Series`,
+        `np.array` or `list`) it will be repeated `obj.shape[0]` times in order to
+        create an array like object from it and then apply the `Series.mask`.
+
+    Returns
+    -------
+    Series
+        Series with the corresponding values based on the conditions, their values
+        and the default value.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:37]: I think the result should be guaranteed to have the same index as `obj`. I believe the current implementation doesn't do this - and if that's intentional then it would be useful to see an example.

**Review by rhshadrach — COMMENTED** [review]
I think this is looking good, some comments.

File: pandas/core/case_when.py (hunk 5 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py:25]: Correct. I added an example in the docstring: 

https://github.com/ELHoussineT/pandas/blob/3e6dc997d7b0a9f0c4b20763306f6003fcc559b1/pandas/core/case_when.py#L62-L71

File: pandas/core/case_when.py (hunk 6 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py:38]: Ah nice catch. Upon looking into this, it turned out we need to maintain the index otherwise some cases will not function as desired (see example below). I fixed it in [cdec601](https://github.com/pandas-dev/pandas/pull/50343/commits/cdec60126dd8e7b3df9b6c7c54ffc47cfdfc9a43). 

Let me know if this does not make "design" sense. 


**Why we need to maintain the index of `obj`?**
Because if we don't, the API cannot support a use case where `obj` is a `Dataframe` and `value0` is a `Series`. To illustrate: 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]), index=['index 1', 'index 2', 'index 3'])
>>> df
         a  b
index 1  1  4
index 2  2  5
index 3  3  6
>>> pd.case_when(
...     df,
...     lambda x: (x.a == 1) & (x.b == 4),
...     df.b,
...     default=0,
... )
0   NaN
1   NaN
2   NaN
dtype: float64
```

why we get that weird series with `NaN`s instead of the expected series? 

Because the index of the `Series` we apply `mask` on does not match the index of `replacements` and `conditions` parameters passed to `mask` function. 

In simply words, notice how `A` does not match `B` and `C` in the picture below: 
image

Comment (ELHoussineT) [reply pandas/core/case_when.py:39]: I realized that with the logic in my comment above, you can still "break" the API as follows: 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]), index=['index 1', 'index 2', 'index 3'])
>>> df
         a  b
index 1  1  4
index 2  2  5
index 3  3  6
>>> pd.case_when(
...     df,
...     lambda x: (x.a == 1) & (x.b == 4),
...     pd.Series([-1,-1,-1], index=['other index 1', 'other index 1', 'other index 1']),,
...     default=0,
... )
index 1    NaN
index 2    0.0
index 3    0.0
dtype: float64
```

With that being said, instead of adding `assert`s to check for matching indices before calling `Series.mask`, I opted for the simple solution of resetting the index for now. See [df1845d](https://github.com/pandas-dev/pandas/pull/50343/commits/df1845d6931293070ea1acbb43697351897f2381). 

Like this, the following use case will not be supported out of the box with this API and the user will need to manually handle the indices (which I think is fine at least for the first iteration of this API): 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]), index=['index 1', 'index 2', 'index 3'])
>>> df
         a  b
index 1  1  4
index 2  2  5
index 3  3  6
>>> df.assign(
...     new_col = case_when(
...         df,
...         lambda x: (x.a == 1) & (x.b == 4),
...         df.b,
...         default=0,
...     )
... )
         a  b  new_col
index 1  1  4      NaN
index 2  2  5      NaN
index 3  3  6      NaN
```

**Comment by ELHoussineT** [comment]
@rhshadrach Answering your question [here](https://github.com/pandas-dev/pandas/pull/50343#discussion_r1131657169) made me realize there the testing should be a bit extensive here as a lot of combos can face this API. To illustrate: 




Anyway, once you agree on the logic, I will proceed to testing.

**Comment by ELHoussineT** [comment]
@rhshadrach How is it looking, shall I proceed and write the tests and docs to wrap up this PR?

File: pandas/core/case_when.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            "The number of conditions and values do not match. "
+            f"There are {len_args - len_args//2} conditions "
+            f"and {len_args//2} values."
+        )
+
+    # construct series on which we will apply `Series.mask`
+    if is_array_like(default):
+        series = pd.Series(default).reset_index(drop=True)
+    else:
+        series = pd.Series([default] * obj.shape[0])
```
Comment (rhshadrach) [inline pandas/core/case_when.py:145]: Instead of reseting the index, can you do `pd.Series(default, index=self.index)` and `pd.Series(default, index=self.index)` for each of these blocks respectively. Then I think you can remove the resetting of the index below.

Comment (rhshadrach) [reply pandas/core/case_when.py:145]: Why use is_array_like over is_list_like?

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/case_when.py (hunk 7 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: > Why use is_array_like over is_list_like?

You are right, updated. [fd12a30](https://github.com/pandas-dev/pandas/pull/50343/commits/fd12a30cb2b40f5643a95bb09c25dfe24466a0da)

Comment (ELHoussineT) [reply pandas/core/case_when.py]: > Instead of reseting the index, can you do pd.Series(default, index=self.index) . This can be done in the else clause as well. Then I think you can remove the resetting of the index below.

If we do so (see b897b8e), then the following case will have an undesired behavior as illustrated in this example: 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]), index=['index 1', 'index 2', 'index 3'])
>>> df
         a  b
index 1  1  4
index 2  2  5
index 3  3  6
>>> pd.case_when(
...     df,
...     lambda x: (x.a == 1) & (x.b == 4),
...     pd.Series([-1,-1,-1], index=['other index 1', 'other index 1', 'other index 1']),
...     default=0,
... )
index 1    NaN
index 2    0.0
index 3    0.0
dtype: float64
```

As a result, I overwrote the index and produced a warning to not shock the user with this index overriding. See [d01d6e8](https://github.com/pandas-dev/pandas/pull/50343/commits/d01d6e897e0a567f9d9c8e3f4bc5fc3aa18d6fe8). This behavior is similar to what was done in other places in Pandas. For example: 

https://github.com/ELHoussineT/pandas/blob/d01d6e897e0a567f9d9c8e3f4bc5fc3aa18d6fe8/pandas/core/frame.py#L3800-L3805

Comment (ELHoussineT) [reply pandas/core/case_when.py]: If the logic looks good, kindly indicate so as I will proceed then to write the tests and docs. Otherwise, we can also continue to iterate on the logic.

Comment (rhshadrach) [reply pandas/core/case_when.py]: > This behavior is similar to what was done in other places in Pandas.

The behavior you're highlighting here is pandas aligning the index to `self`'s index. The logic in this PR entirely ignores the index. I would expect pandas to align the index, so that the NaN value in the first row is expected.

Comment (ELHoussineT) [reply pandas/core/case_when.py]: Hmm.. I see. What do you suggest in this case?

File: pandas/core/case_when.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+            "The number of conditions and values do not match. "
+            f"There are {len_args - len_args//2} conditions "
+            f"and {len_args//2} values."
+        )
+
+    # construct series on which we will apply `Series.mask`
+    if is_list_like(default):
+        series = pd.Series(default.values, index=obj.index)
+    else:
+        series = pd.Series([default] * obj.shape[0], index=obj.index)
```
Comment (rhshadrach) [inline pandas/core/case_when.py:157]: I think use `pd.Series(default, index=obj.index)` in both cases here.

File: pandas/core/case_when.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+            obj.index
+        ):
+            replacements = warn_and_override_index(
+                replacements, f"(in args[{i+1}])", obj.index
+            )
+
+        if isinstance(conditions, pd.Series) and not conditions.index.equals(obj.index):
+            conditions = warn_and_override_index(
+                conditions, f"(in args[{i}])", obj.index
+            )
```
Comment (rhshadrach) [inline pandas/core/case_when.py:179]: Remove this entirely, mask will align.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/case_when.py (hunk 9 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: Sure, updated. da79aa653b37164035520cc073f37c60ba5dca13

File: pandas/core/case_when.py (hunk 8 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: Makes sense. Updated [eb70eaf](https://github.com/pandas-dev/pandas/pull/50343/commits/eb70eafcd549341ce30ebfad90f14e6f95fc77d3).

File: pandas/core/case_when.py (hunk 9 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: Please take another look, if it looks good, I will proceed to tests and docs. Thanks.

**Comment by ELHoussineT** [comment]
@rhshadrach PTAL.

File: pandas/core/case_when.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        if callable(args[i]):
+            conditions = com.apply_if_callable(args[i], obj)
+        else:
+            conditions = args[i]
+
+        # get replacements
+        replacements = args[i + 1]
+
+        # `Series.mask` call
+        series = series.mask(conditions, replacements)
```
Comment (rhshadrach) [inline pandas/core/case_when.py:165]: I believe we want to only modify the value on the first `conditions` that evaluates to True (correct me if you think this is wrong!); this will modify it for all conditions. One approach is to maintain a `modified` that starts as `pd.Series(False, index=self.index)`. Then this line can become `series = series.mask(~modified & conditions, replacements)`. After this line, we also need to update `modified |= conditions`.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/case_when.py (hunk 10 above)
Comment (ELHoussineT) [reply pandas/core/case_when.py]: Good point. Updated ([f2bf4d8](https://github.com/pandas-dev/pandas/pull/50343/commits/f2bf4d866a38fd888561b4c3d2acc537051b7cdc)). 

Now, if multiple conditions are met, the value of the first one is used. For example: 

```
>>> df 
         a  b
index 1  1  4
index 2  2  5
index 3  3  6

>>> pd.case_when(
... df,
... lambda x: x.a > 0,
... 1,
... lambda x: x.a == 1, 
... -1,
... default='default'
)
index 1    1
index 2    1
index 3    1
dtype: object

**Comment by ELHoussineT** [comment]
@rhshadrach I also had to add `.convert_dtypes()` [a5d678f](https://github.com/pandas-dev/pandas/pull/50343/commits/a5d678f067b331c9670ee6de2073f9fd79a3a6b4). 

This is necessary since the default always determines the first dtype of the series even if the final series does not have any default value. 

To illustrate, see [1] which does not have `.convert_dtypes()` while [2] has it. Notice how [2] has the correct dtype in the result series. 

[1] : 
```
>>> df 
         a  b
index 1  1  4
index 2  2  5
index 3  3  6

>>> pd.case_when(
... df,
... lambda x: x.a > 0,
... 1,
... lambda x: x.a == 1, 
... -1,
... default='default'
)
index 1    1
index 2    1
index 3    1
dtype: object

```

[2] : 

```
>>> df 
         a  b
index 1  1  4
index 2  2  5
index 3  3  6

>>> pd.case_when(
... df,
... lambda x: x.a > 0,
... 1,
... lambda x: x.a == 1, 
... -1,
... default='default'
)
index 1    1
index 2    1
index 3    1
dtype: Int64
```

**Comment by rhshadrach** [comment]
> @rhshadrach I also had to add `.convert_dtypes()` [a5d678f](https://github.com/pandas-dev/pandas/pull/50343/commits/a5d678f067b331c9670ee6de2073f9fd79a3a6b4).
> 
> This is necessary since the default always determines the first dtype of the series even if the final series does not have any default value.

Thanks for identifying this, it is indeed something I had not considered. I do not think Int64 is the right answer in the example you gave. As values of the input change, the dtype of the result can change, which can then lead to radically different answers later on. It is an example of "values-dependent behavior" which we try to avoid if at all possible.

I think we should be using something like `find_common_type` to determine what the output dtype is based on the input dtypes alone (and not values!). However I don't know if this is readily used since this function accepts scalars. In short, I'm not sure of what a good solution looks like here.

**Comment by ELHoussineT** [comment]
@rhshadrach 

> I do not think Int64 is the right answer in the example you gave.

May I ask why? It is still not super clear to me why can't we proceed with `.convert_dtypes()`. 

Thanks

**Comment by rhshadrach** [comment]
> May I ask why? 

I included why I think to be the case in my initial response.

**Comment by ELHoussineT** [comment]
@rhshadrach 

Understood and I agree. 

In this case, I would suggest to remove `.convert_dtypes()` (see [072bc0c](https://github.com/pandas-dev/pandas/pull/50343/commits/072bc0c648a11fcb9fcb27b6dd3276a9a3997b59)) and mention this limitation in the docstring (see [747b379](https://github.com/pandas-dev/pandas/pull/50343/commits/747b379070f2bbcae3b5ca24e4306ffdc7a919b0)). 

The reason why I suggested the above is that this issue is built in `Series.mask()` and keeping the behavior consistent for now might be beneficial. 

To illustrate how the issue is in `Series.mask()`, see the dtype of the result `Series` below: 

```
>>> s = pd.Series(['a','b'])
0    a
1    b
dtype: object

>>> s.mask([True, True], 1)
0    1
1    1
dtype: object
```

Keeping that behavior as is in `pd.case_when` will maintain the consistency (even through one could argue it is an undesired behavior) and if this gets fixed in `Series.mask` the changes will propagate to `pd.case_when` as it uses `Series.mask` under the hood. 

I am also happy to tackle this issue in `Series.mask` after we are done with this PR. I am optimistic that we can find a solution. 

Looking forward to your reply. If you agree with the above, please indicate that so I proceed with tests & docs. 

Best,

**Comment by ELHoussineT** [comment]
@rhshadrach FYI https://github.com/pandas-dev/pandas/issues/52662

**Comment by ELHoussineT** [comment]
@rhshadrach PTAL.

File: doc/source/whatsnew/v2.0.0.rst [hunk 11]
```diff
@@ -14,6 +14,26 @@ including other versions of pandas.
 Enhancements
 ~~~~~~~~~~~~
 
+.. _whatsnew_200.enhancements.case_when:
+
+Assignment based on multiple conditions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``pd.case_when`` API has now been added to support assignment based on multiple conditions.
```
Comment (rhshadrach) [inline doc/source/whatsnew/v2.0.0.rst:22]: Add a proper link to the API documentation: `` :func:`case_when` ``

Also, can you refer to this as a function rather than an API.

File: pandas/core/case_when.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
+from __future__ import annotations
+
+from typing import Any
+import warnings
+
+from pandas.util._exceptions import find_stack_level
+
+import pandas as pd
```
Comment (rhshadrach) [inline pandas/core/case_when.py:8]: Don't import all of pandas, only what you need from various modules.

File: pandas/core/case_when.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        f"Series {series_type} will be reindexed to match obj index.",
+        UserWarning,
+        stacklevel=find_stack_level(),
+    )
+    return pd.Series(series.values, index=index)
+
+
+def case_when(obj: pd.DataFrame | pd.Series, *args, default: Any) -> pd.Series:
+    """
+    Returns a Series based on multiple conditions assignment.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:23]: Can we say "Construct" instead of Returns. Also, "multiple conditions assignment" sounds off to me, I would recommend just "multiple conditions".

File: pandas/core/case_when.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+    This is useful when you want to assign a column based on multiple conditions.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:25]: This can be used independently of assigning a column, I suggest this be removed.

File: pandas/core/case_when.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+    Uses `Series.mask` to perform the assignment.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:26]: I think this is an implementation detail.

File: pandas/core/case_when.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+    If multiple conditions are met, the value of the first condition is taken.
+
+    The returned Series have the same index as `obj` as mainly follows dtype
+    of `default`.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:31]: What does "as mainly follows dtype of `default`" mean here?

File: pandas/core/case_when.py [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+    Parameters
+    ----------
+    obj : Dataframe or Series on which the conditions will be applied.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:35]: The first line should just be the dtype; a description of the argument should appear on the subsequent line.

File: pandas/core/case_when.py [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+    args : Variable argument of conditions and expected values.
+        Takes the form:
+            `condition0`, `value0`, `condition1`, `value1`, ...
+        `condition` can be a 1-D boolean array/series or a callable
+        that evaluate to a 1-D boolean array/series. See examples below.
+    default : Any
+        The default value to be used if all conditions evaluate False. This value
+        will be used to create the `Series` on which `Series.mask` will be called.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:43]: Series.mask is an implementation detail.

File: pandas/core/case_when.py [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+        If this value is not already an array like (i.e. it is not of type `Series`,
+        `np.array` or `list`) it will be repeated `obj.shape[0]` times in order to
+        create an array like object from it and then apply the `Series.mask`.
```
Comment (rhshadrach) [inline pandas/core/case_when.py:46]: Same here

File: pandas/core/case_when.py [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+    # construct series on which we will apply `Series.mask`
+    series = pd.Series(default, index=obj.index)
+
+    # a series to keep track of which row got modified.
+    # we need this because if a row satisfied multiple conditions,
+    # we set the value of the first condition.
+    modified = pd.Series(False, index=obj.index)
+
+    for i in range(0, len_args, 2):
+        # get conditions
```
Comment (rhshadrach) [inline pandas/core/case_when.py:163]: This comment is essentially repeating the code; can you remove.

File: pandas/core/case_when.py [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+        if callable(args[i]):
+            conditions = com.apply_if_callable(args[i], obj)
+        else:
+            conditions = args[i]
+
+        # get replacements
```
Comment (rhshadrach) [inline pandas/core/case_when.py:169]: Same.

File: pandas/core/case_when.py [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+        replacements = args[i + 1]
+
+        # `Series.mask` call
```
Comment (rhshadrach) [inline pandas/core/case_when.py:172]: Same

File: pandas/core/case_when.py [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+        series = series.mask(~modified & conditions, replacements)
+
+        # keeping track of which row got modified
```
Comment (rhshadrach) [inline pandas/core/case_when.py:175]: Same

File: pandas/tests/test_case_when.py [hunk 24]
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+import pytest  # noqa
```
Comment (rhshadrach) [inline pandas/tests/test_case_when.py:2]: Why is this needed? Include the error code when using noqa.

File: pandas/tests/test_case_when.py [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+
+import pandas as pd
+import pandas._testing as tm
+
+
+class TestCaseWhen:
```
Comment (rhshadrach) [inline pandas/tests/test_case_when.py:8]: If not making use of the class, then use test functions instead of methods.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
The implementation looks good, but this needs many additional tests. In particular, we should be testing the behavior when invalid inputs are given.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
